File: ios/Classes/ObserverStatus.swift
//
//  LastDateSavedHandler.swift
//  health_metrics_observers
//
//  Created by Sudarshan Chakra on 14/03/22.
//

import Foundation

class ObserverStatus {
    static let defaults = UserDefaults.standard
    
    //observer
    static let observerSyncingKey = "observerSyncing"
    static let observerCreatedKey = "observerCreated"
    //steps
    static let nextStartDateKey = "nextStartDate"
    static let lastStepCountSentKey = "lastStepCountSent"
    static let lastAttemptToSendKey = "lastAttemptToSend"
    static let observerOnlyLastSavedKey = "lastDateSavedObserver"
    //sleep
    static let lastSleepDateSavedKey = "lastSleepDateSaved"
    static let lastSleepValueSentKey = "lastSleepValueSent"
    
    static let apiUrlKey = "apiUrl"
    static let migratedToShrcrApiKey = "migratedToSharecareApi"
    
    static let lastCheckWorkoutsKey = "lastCheckWorkouts"
    
    // last saved date - shared across sources
    static func getLastDateSaved(metricType: String) -> Int? {
        let lastDateSavedKey = Self.getLastDateSavedKeyByMetricType(metricType: metricType)
        if let nextStartDate = defaults.object(forKey: lastDateSavedKey) as? Date {
            return Int(nextStartDate.timeIntervalSince1970 * 1000)
        } else {
            return nil
        }
    }
    
    static func getLastReferenceDate(metricType: MetricType) -> Date? {
        //reference date last check or saved
        let lastDateSavedKey = Self.getLastDateSavedKeyByMetricType(metricType: metricType.rawValue)
        return defaults.object(forKey: lastDateSavedKey) as? Date
    }
    
    static func updateLastDateSaved(newDateInMilliseconds: Int, metricType: String){
        let dateInSeconds = Double(newDateInMilliseconds/1000)
        let newDate = Date(timeIntervalSince1970: dateInSeconds)
        let lastDateSavedKey = Self.getLastDateSavedKeyByMetricType(metricType: metricType)
        defaults.set(newDate, forKey: lastDateSavedKey)
    }
    
    static func getLastDateSavedKeyByMetricType(metricType: String) -> String {
        switch (metricType) {
        case "STEP":
            return Self.nextStartDateKey
        case "SLEEP":
            return Self.lastSleepDateSavedKey
        case "WORKOUT":
            return Self.lastCheckWorkoutsKey
        default:
            return ""
        }
    }
    
    static func updateObserverSyncing(newValue: Bool) {
        defaults.set(newValue, forKey: observerSyncingKey)
    }
    
    static func isObserverSyncing() -> Bool {
       return defaults.bool(forKey: observerSyncingKey)
    }
    
    // last saved date - observer only
    static func getLastDateSavedObserverOnly() -> Int? {
        if let nextStartDate = defaults.object(forKey: Self.observerOnlyLastSavedKey) as? Date {
            return Int(nextStartDate.timeIntervalSince1970 * 1000)
        } else {
            return nil
        }
    }
    
    static func getLastAttemptDate() -> Int? {
        if let lastAttemptDate = defaults.object(forKey: Self.lastAttemptToSendKey) as? Date {
            return Int(lastAttemptDate.timeIntervalSince1970 * 1000)
        } else {
            return nil
        }
    }

    static func updateObserverCreated(_ created: Bool, metricType: MetricType) {
        let createdKey = metricType.getObserverCreatedKeyStr()
        defaults.set(created, forKey: createdKey)
    }
    
    static func hasObserverCreated(metricType: MetricType) -> Bool {
        let createdKey = metricType.getObserverCreatedKeyStr()
        return defaults.bool(forKey: createdKey)
    }
    
    static func updateApiUrl(_ newApiUrl: String){
        defaults.set(newApiUrl, forKey: Self.apiUrlKey)
    }
    
    static func getApiUrl() -> String? {
        return defaults.string(forKey: Self.apiUrlKey)
    }
    
    static func updateMigrated(_ migrated: Bool){
        defaults.set(migrated, forKey: Self.migratedToShrcrApiKey)
    }
    
    static func getMigrated() -> Bool {
        return defaults.bool(forKey: Self.migratedToShrcrApiKey)
    }
    
    static func getObserverStatus() -> [String: Any?] {
        let lastDateSavedSharedStep = Self.getLastDateSaved(metricType: "STEP")
        let lastDateSavedSharedSleep = Self.getLastDateSaved(metricType: "SLEEP")
        return [
            "last_saved": lastDateSavedSharedStep,
            "last_saved_date_across_sources": lastDateSavedSharedStep,
            "last_saved_date_observer_only": Self.getLastDateSavedObserverOnly(),
            "last_steps_count_saved": defaults.integer(forKey: Self.lastStepCountSentKey),
            "last_attempt_timestamp": Self.getLastAttemptDate(),
            "created": Self.hasObserverCreated(metricType: .step),
            "observer_syncing": Self.isObserverSyncing(),
            "last_saved_sleep_date_across_sources": lastDateSavedSharedSleep,
            "workouts_observer_created": Self.hasObserverCreated(metricType: .workout),
            "api_url": Self.getApiUrl(),
            "migrated_to_shrcr_api": Self.getMigrated()
        ]
    }
    
}

File: android/src/main/kotlin/com/wecare/health_metrics_observers/ObserverStatus.kt
package co.sinapsis.sinapsis_health_metrics

import android.content.Context
import java.text.SimpleDateFormat
import java.util.*

class AppConstants {
    companion object {
        const val defaultDateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ"
        const val sharedPrefId = "wecare.default.preferences"
    }
}

class ObserverStatus {
    companion object {
        //observer
        const val observerCreatedKey = "observer_created"
        const val observerSyncingKey = "observer_syncing"
        //steps
        const val lastAttemptToSendKey = "last_attempt_to_send"
        const val observerOnlyLastDateSavedKey = "last_date_saved_observer"
        const val lastStepCountSentKey = "last_step_count_sent"
        const val lastDateSavedKey = "last_date_saved"
        //sleep
        const val lastSleepValueSentKey = "last_sleep_value_sent"
        const val lastSleepDateSavedKey = "last_sleep_date_saved"


        fun getLastDateSaved(context: Context, metricType: String): Long? {
            val sharedPref =
                context.getSharedPreferences(AppConstants.sharedPrefId, Context.MODE_PRIVATE)
            val dateFormat = SimpleDateFormat(AppConstants.defaultDateFormat)
            val lastDateSavedKeyByType = getLastDateSavedKeyByMetricType(metricType)
            val dateSavedString = sharedPref.getString(lastDateSavedKeyByType, null)

            if (dateSavedString != null) {
                val lastDateSaved = dateFormat.parse(dateSavedString)
                return lastDateSaved?.time ?: null
            }else{
                return null
            }
        }

        fun updateLastDateSaved(context: Context, newDateInMilliseconds: Long, metricType: String) {
            val sharedPref = context.getSharedPreferences(AppConstants.sharedPrefId, Context.MODE_PRIVATE)
            val dateFormat = SimpleDateFormat(AppConstants.defaultDateFormat)
            val newDate = Date(newDateInMilliseconds)
            val lastDateSavedKeyByType = getLastDateSavedKeyByMetricType(metricType)
            with (sharedPref.edit()) {
                putString(lastDateSavedKeyByType, dateFormat.format(newDate))
                apply()
            }
        }

        private fun getLastDateSavedKeyByMetricType(metricType: String): String {
            return when (metricType) {
                "STEP" -> lastDateSavedKey
                "SLEEP" -> lastSleepDateSavedKey
                else -> ""
            }
        }

        private fun getLastAttemptDate(context: Context): Long? {
            val sharedPref =
                context.getSharedPreferences(AppConstants.sharedPrefId, Context.MODE_PRIVATE)
            val dateFormat = SimpleDateFormat(AppConstants.defaultDateFormat)
            val dateSavedString = sharedPref.getString(lastAttemptToSendKey, null)
            if (dateSavedString != null) {
                val lastAttempt = dateFormat.parse(dateSavedString)
                return lastAttempt?.time ?: null
            }else{
                return null
            }
        }

        //last saved - observer only
        fun getLastDateSavedObserverOnly(context: Context): Long? {
            val sharedPref =
                context.getSharedPreferences(AppConstants.sharedPrefId, Context.MODE_PRIVATE)
            val dateFormat = SimpleDateFormat(AppConstants.defaultDateFormat)
            val dateSavedString = sharedPref.getString(observerOnlyLastDateSavedKey, null)

            if (dateSavedString != null) {
                val lastDateSaved = dateFormat.parse(dateSavedString)
                return lastDateSaved?.time ?: null
            }else{
                return null
            }
        }

        fun isObserverSyncing(context: Context) : Boolean {
            val sharedPref = context.getSharedPreferences(AppConstants.sharedPrefId, Context.MODE_PRIVATE)
            return sharedPref.getBoolean(observerSyncingKey, false)
        }

        fun updateObserverSyncing(context: Context, newValue: Boolean){
            val sharedPref = context.getSharedPreferences(AppConstants.sharedPrefId, Context.MODE_PRIVATE)
            with (sharedPref.edit()) {
                putBoolean(observerSyncingKey, newValue)
                apply()
            }
        }

        fun updateObserverCreated(context: Context, created: Boolean){
            val sharedPref =
                context.getSharedPreferences(AppConstants.sharedPrefId, Context.MODE_PRIVATE)
            with (sharedPref.edit()) {
                putBoolean(observerCreatedKey, created)
                apply()
            }
        }

        fun hasObserverCreated(context: Context): Boolean {
            val sharedPref =
                context.getSharedPreferences(AppConstants.sharedPrefId, Context.MODE_PRIVATE)
            return sharedPref.getBoolean(observerCreatedKey, false)
        }

        fun getStatusMap(context: Context): Map<String, Any?> {
            val sharedPref =
                context.getSharedPreferences(AppConstants.sharedPrefId, Context.MODE_PRIVATE)
            val lastDateSavedSharedSteps = ObserverStatus.getLastDateSaved(context, "STEP")
            val lastDateSavedSharedSleep = ObserverStatus.getLastDateSaved(context, "SLEEP")
            return mapOf<String, Any?>(
                "last_saved" to lastDateSavedSharedSteps,
                "last_saved_date_across_sources" to lastDateSavedSharedSteps,
                "last_saved_date_observer_only" to ObserverStatus.getLastDateSavedObserverOnly(context),
                "last_attempt_timestamp" to ObserverStatus.getLastAttemptDate(context),
                "last_steps_count_saved" to sharedPref.getInt(lastStepCountSentKey, 0),
                "observer_syncing" to isObserverSyncing(context),
                "created" to ObserverStatus.hasObserverCreated(context),
                "last_saved_sleep_date_across_sources" to lastDateSavedSharedSleep,
            )
        }
    }
}